anesthetic/weighted_pandas.py

```diff
@@ -2,31 +2,31 @@
 
 import numpy as np
 import pandas
-from warnings import warn
 from anesthetic.utils import compress_weights, channel_capacity, quantile
 
 
 class _WeightedObject(object):
     @property
     def weight(self):
         """Sample weights."""
-        if self._weight is None:
+        if self.index.nlevels == 1:
             return pandas.Series(index=self.index, data=1.)
         else:
-            return self._weight[self.index]
+            return pandas.Series(data=self.index.droplevel(0),
+                                 index=self.index)
+
+    @weight.setter
+    def weight(self, weight):
+        if weight is not None:
+            self.index = [self.index.get_level_values(0), weight]
+            self.index.set_names([None, 'weight'], inplace=True)
 
     @property
     def _rand(self):
         """Random number for consistent compression."""
-        return self._rand_[self.index]
-
-    def _construct_weights(self, weight):
-        if weight is not None:
-            self._weight = pandas.Series(index=self.index, data=weight)
-        else:
-            self._weight = None
-        rand = np.random.rand(len(self))
-        self._rand_ = pandas.Series(index=self.index, data=rand)
+        h = pandas.util.hash_array(self.index)
+        _rand = np.frexp(h)[0]*2-1
+        return pandas.Series(data=_rand, index=self.index)
```

Comment:
> One final slightly hacky thing that came with this was an update to the _rand functionality, which is now deterministically generated from the MultiIndex. I couldn't think of a neater way to do this post pandas 1.1, but it is something to watch out for.

I'm unfamiliar with the `np.frexp` function, does this work like a random number generator? Could we alternatively use the pandas hash as seed for a random number generator?